cpp/oneapi/dal/backend/primitives/selection/test/perf_select_flagged_dpc.cpp

```diff
@@ -0,0 +1,291 @@
+#include "oneapi/dal/test/engine/io.hpp"
+#include "oneapi/dal/test/engine/math.hpp"
+#include "oneapi/dal/backend/primitives/selection/select_flagged.hpp"
+
+namespace oneapi::dal::backend::primitives::test {
+
+namespace te = dal::test::engine;
+namespace la = te::linalg;
+namespace de = dal::detail;
+
+template <typename TestType>
+class select_flagged_test : public te::policy_fixture {
+public:
+    using Float = std::tuple_element_t<0, TestType>;
+    using Flag = std::tuple_element_t<1, TestType>;
+    using Integer = std::uint32_t;
+
+    auto allocate_arrays(Integer elem_count) {
+        auto& q = this->get_queue();
+        auto val = ndarray<Float, 1>::empty(q, { elem_count }, sycl::usm::alloc::device);
+        auto mask = ndarray<Flag, 1>::empty(q, { elem_count }, sycl::usm::alloc::device);
+
+        return std::make_tuple(val, mask);
+    }
+
+    auto allocate_vector_of_arrays(std::int64_t vector_count, Integer elem_count) {
+        auto& q = this->get_queue();
+        std::vector<ndarray<Float, 1>> val_vec(vector_count);
+        std::vector<ndarray<Flag, 1>> mask_vec(vector_count);
+        for (std::int64_t i = 0; i < vector_count; i++) {
+            val_vec[i] = ndarray<Float, 1>::empty(q, { elem_count }, sycl::usm::alloc::device);
+            mask_vec[i] = ndarray<Flag, 1>::empty(q, { elem_count }, sycl::usm::alloc::device);
+        }
+
+        return std::make_tuple(val_vec, mask_vec);
+    }
+
+    auto init_vector_of_arrays(std::vector<ndarray<Float, 1>>& val_vec,
+                               std::vector<ndarray<Flag, 1>>& mask_vec,
+                               Float a,
+                               Float b,
+                               std::int64_t seed = 777) {
+        const std::int64_t vector_count = val_vec.size();
+        for (std::int64_t i = 0; i < vector_count; i++) {
+            fill_uniform(val_vec[i], mask_vec[i], a, b, seed);
+        }
+
+        return std::make_tuple(val_vec, mask_vec);
+    }
+
+    void fill_uniform(ndarray<Float, 1>& val,
+                      ndarray<Flag, 1>& mask,
+                      Float a,
+                      Float b,
+                      std::int64_t seed = 777) {
+        Integer elem_count = de::integral_cast<Integer>(val.get_count());
+        std::mt19937 rng(seed);
+        std::uniform_real_distribution<Float> distr(a, b);
+
+        Float pivot = distr(rng);
+        // move generation to device when rng is available there
+        auto val_host = ndarray<Float, 1>::empty({ elem_count });
+        auto mask_host = ndarray<Flag, 1>::empty({ elem_count });
+        Float* val_ptr = val_host.get_mutable_data();
+        Flag* mask_ptr = mask_host.get_mutable_data();
+
+        for (Integer el = 0; el < elem_count; el++) {
+            val_ptr[el] = distr(rng);
+            mask_ptr[el] = val_ptr[el] < pivot ? 1 : 0;
+        }
+
+        auto& q = this->get_queue();
+
+        val = val_host.to_device(q);
+        mask = mask_host.to_device(q);
```

Comment:
It looks like, there is no need for separation on allocation and initialization any more. You can do the following:
```c++
TEST(...) {
    auto [val, mask] = this->init_uniform(elem_count, -25., 25.);
}
``` 
instead
```c++
TEST(...) {
    auto [val, mask] = this->allocate_arrays(elem_count);
    this->fill_uniform(val, mask, -25., 25.);
}
```

Reply:
The same for `init_vector_of_arrays` memory allocation and initialization can be done in the same loop.

Reply:
replaced to_device with val.assign(q, val_host);